WordPress/Classes/Services/BloggingPromptsService.swift

```diff
@@ -330,17 +330,13 @@ private extension BloggingPromptsService {
     /// - Parameters:
     ///   - remoteSettings: The blogging prompt settings from the remote.
     ///   - completion: Closure to be called on completion.
-    func saveSettings(_ remoteSettings: RemoteBloggingPromptsSettings, completion: @escaping () -> Void) {
-        let derivedContext = contextManager.newDerivedContext()
-        derivedContext.perform {
-            let settings = self.loadSettings(context: derivedContext) ?? BloggingPromptSettings(context: derivedContext)
-            settings.configure(with: remoteSettings, siteID: self.siteID.int32Value, context: derivedContext)
-
-            self.contextManager.save(derivedContext) {
-                DispatchQueue.main.async {
```

Comment:
This is an unfortunate API design that I'm not sure we should bake into the new method signature – it's never a good idea to make assumptions about where a callback will happen (if we chain this with other methods, it'll non-deterministically consume main thread time, which can make it nearly impossible to maintain scrolling/animation performance).

While a slightly better method signature could be something like:

```swift
func saveSettings(_ remoteSettings: RemoteBloggingPromptsSettings, completion: @escaping () -> Void, queue: DispatchQueue = .main) {
```

it doesn't actually solve our problem, which is that when using `Task`, the only safe way to execute a callback on on the `MainActor`.

IMHO we should really retain the callback-based approach here, and if/when we're ready to adopt async/await, we add new methods alongside callback-based ones and use those from the UI layer on down. I'm just not sure that it's possible to safely mix callbacks and async/await in safe _and_ performant way?

Reply:
> it's never a good idea to make assumptions about where a callback will happen

+1. Unless the callback _has to_ run on a particular thread. That's usually the case for UI code that needs to run on the main thread, but should not be the case here, in "service" code.

>  our problem, which is that when using `Task`, the only safe way to execute a callback is on on the `MainActor`.

@jkmassel the way I'm reading "safe" is "be guaranteed it runs on the main thread". My understanding is that we can run callback from a `Task` just by calling (in case of a closure `callback()`) them. But that's "not safe" because there's no guarantee on which actor / thread they'll run.

I think modern Swift allows caller to use the `@MainActor` annotation to tell the build system to run async code on the main thread. Unfortunately, we find ourselves in a mixed environment with different layers of legacy code from different times. Because of that, I think the suggestion of injecting the `DispatchQueue` in this method's signature is a good, small step to help us bridge the gap.

> I'm just not sure that it's possible to safely mix callbacks and async/await in safe and performant way?

Does "safely" here refer to thread safety, or "so that they run on the thread that we want / would expect"? Also, what is the performance concern? _Hopefully we'll have better performance monitoring via Sentry soon._ 🙏

Reply:
> . Because of that, I think the suggestion of injecting the `DispatchQueue` in this method's signature is a good, small step to help us bridge the gap.

On the other hand... If a consumer has a specific requirement on the thread a callback should run, should it be their responsibility to specify that thread, because of the "make no assumptions rationale"?

Is passing a queue as part of the method signature better or worse than implicitly expecting callers to do something like

```swift
service.saveSettings(settings) {
  DispatchQueue.main.async {
    // completion code I want to run on the main thread?
  }
}
```

The upside of the "make no assumption" approach is that it's clean. The downside is that it might be easy to forget about the main thread.

Maybe an alternative is to require the `DispatchQueue` parameter, but not provide a default value, so that every caller is forced to think about threading.

Oh... Async is hard 🙃

Reply:
IMO, whether a closure argument is going to be called from the main thread or a background thread should be part of the API, either by documenting it or making it part of the function signature. A couple of examples like this are `URLSession` and `NSManagedContextObject`. So I agree with the "make no assumption" rule, since the API should make it explicit.

Reply:
@mokagio – just to address your question:

> Does "safely" here refer to thread safety, or "so that they run on the thread that we want / would expect"? Also, what is the performance concern? Hopefully we'll have better performance monitoring via Sentry soon. 🙏

Thread safety, in this case. The only "certain" thread is the main one (and as you noted, the thread that services `@MainActor`). But blocking the main thread to sync background threads will have terrible performance implications, which was why I noted it.